File: sem2/cpp/cll.cpp
#include <bits/stdc++.h>

using namespace std;

class node {
public:
	int data;
	node* next;
};

node* makehead(int data) {
	node* root = new node();
	root->next = root;
	root->data = data;
	return root;
}

void printlist(node* head) {
	cout << head->data << " ";
	node* tmp = head->next;
	while(tmp!=head) {
		cout << tmp->data << " ";
		tmp = tmp->next;
	}
}

void addnode(node* head, int data) {
	node* tmp = head;
	while(tmp->next != head)
		tmp = tmp->next;
	node* cur = new node();
	cur->data = data;
	tmp->next = cur;
	cur->next = head;
}

void reverse(node* &head) {
	node* prev = head, *tmp = head->next;
	while(tmp != head) {
		node* cur = tmp->next;
		tmp->next = prev;
		prev = tmp;
		tmp = cur;
	}
	head->next = prev;
	head = prev;
}

void deletenode(node* &head, int index) {
	int cnt = 1; node* tmp = head;
	if(index == 1)	{
		tmp = head->next;
		while(tmp->next != head)
			tmp = tmp->next;
		tmp->next = head->next;
		head->next = NULL;
		node* cur = head;
		head = tmp->next;
		delete cur;
	}
	else {
		while(cnt != index-1) {
			tmp = tmp->next;
			cnt++;
		}
		node* cur = tmp->next;
		tmp->next = cur->next;
		cur->next = NULL;
		delete cur;
	}
}
int main() {
	node* root = makehead(5);
	addnode(root, 4); addnode(root, 3); addnode(root, 2); addnode(root, 1);
	printlist(root); cout << "\n";
	reverse(root);
	deletenode(root, 1);
	printlist(root);
	return 0;
}

File: sem2/python/cll.py
#!/usr/bin/python3

class node:
    def __init__(self,data):
        self.data = data
        self.next = None

class circular_list:
    def __init__(self):
        self.head = None
    def makehead(self, data):
        self.head = node(data)
        self.head.next = self.head
    def addnode(self, data):
        tmp = self.head
        while(tmp.next != self.head):
            tmp = tmp.next
        tmp.next = node(data)
        cur = tmp.next
        cur.next = self.head
    def printlist(self):
        if(self.head):
            print(self.head.data, end = " ")
        tmp = self.head.next
        while(tmp != self.head):
            print(tmp.data, end = " ")
            tmp = tmp.next
    def deletenode(self, index):
        tmp = self.head
        if(index == 1):
            while(tmp.next != self.head):
                tmp = tmp.next
            head = tmp.next
            tmp.next = head.next
            self.head = tmp.next
            head.next = None
        else:
            cnt = 1
            if(index > 2):
                cnt += 1
                tmp = tmp.next
                while(tmp != self.head and cnt < index-1):
                    tmp = tmp.next
                    cnt+=1
            cur = tmp.next
            tmp.next = cur.next
            cur.next = None
    def reverse(self):
        prev = self.head
        tmp = self.head.next
        while(tmp != self.head):
            cur = tmp.next
            tmp.next = prev
            prev = tmp
            tmp = cur
        tmp.next = prev
        self.head = prev

cll = circular_list()
cll.makehead(1)
cll.addnode(2)
cll.addnode(3)
cll.addnode(4)
cll.addnode(5)
cll.addnode(6)
cll.deletenode(6)
cll.printlist()
cll.reverse()
cll.printlist()


        
